File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int max = 0;
        int start = 0; int end = 0;
        Map<Character, Integer> map = new HashMap<>();
        while(end < s.length()){
            if(map.containsKey(s.charAt(end)) && map.get(s.charAt(end)) >= start){
                start = map.get(s.charAt(end)) + 1;
            }
               
            map.put(s.charAt(end), end);
               
            max = Math.max(max, end - start + 1);
            end++;
        }
    
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        sub = ''
        for c in s:
            if c not in sub:
                sub += c
                ans = max(ans, len(sub))
            else:
                cut = sub.index(c)
                sub = sub[cut + 1:] + c
        return ans